### DISCUSSION TITLE
Parsing model Indices in PDB files containing multiple models

### DISCUSSION THREAD

**Discussion by a-r-j** [BioPandas/biopandas#100] [Ideas]
Hi @rasbt!

I'm not sure if this is a missing feature (or if I'm being silly!) but I don't see that PDB files containing multiple models (eg. [2jyf](https://www.rcsb.org/structure/2jyf)) denote the start and end of each model in the PDB file. Granted this is an RNA structure but it is parsed just fine otherwise. It would be nice to add another column to the dataframe (or as another entry in the PandasPDB object) denoting which model each atom in the main coordinate text belongs to.

EDIT:
Ah, I just spotted you're already aware #50

**Comment by a-r-j** [comment]
After musing on the solutions outlined  in #50, I think I would favour functionality like:

`df.label_models()`

Which adds a `"model"` column to the contained ATOM/HETATM etc dataframes.

This then enables users to quickly grab the correct selection of atoms using normal pandas selections

eg something like: 

```python
df = df.df["ATOM"]

model_1 = df.loc[df["model"] == 1]
```

this could then be wrapped in a more general method like:

```python
from biopandas.pdb import PandasPdb

df = PandasPdb.read_pdb("path/to/structure.pdb")

model_1 = df.get_model(index=3)
```

Where `model_1` is another `PandasPdb` instance containing only the atoms that are present in that model.

**Reply by rasbt** [reply]
Yeah, unfortunately, I never got around to add that 😅. Do you think there is an easy way to parse out the model information from the PDB? Something that is not error prone?

> After musing on the solutions outlined in https://github.com/rasbt/biopandas/issues/50, I think I would favour functionality like: 
df.label_models()

Do you mean this is something the user has to manually call after loading the PDB, or would that be an internally used method?

Because regarding your example

```python
from biopandas.pdb import PandasPdb

df = PandasPdb.read_pdb("path/to/structure.pdb")

model_1 = df.get_model(index=3)
```

this would be

```python
from biopandas.pdb import PandasPdb

df = PandasPdb.read_pdb("path/to/structure.pdb")
df.label_models()
model_1 = df.get_model(index=3)
```
?

Otherwise I like your idea of having an additional column. That is probably the most elegant way to handle it.

**Reply by a-r-j** [reply]
Think I figured it out in #101. The model start and end line indices can be grabbed from `df.df["OTHERS"]`, from there it's easy to grab the selections. I added a few different utilities on top of it for getting selections (see the tutorial nb for examples).
